src/main/process-pool.ts

```diff
@@ -0,0 +1,70 @@
+import { ChildProcess, fork } from 'child_process';
+import * as os from 'os';
+import { resolve as pathResolve } from 'path';
+import { removeDebugArgs } from './util';
+
+interface IMochaProcess extends ChildProcess {
+  destroy: () => void;
+}
+
+export default class ProcessPool {
+  private maxParallel: number;
+  private waitingList: Array<(process: IMochaProcess) => void> = [];
+  private unusedProcesses: IMochaProcess[] = [];
+  private processes: IMochaProcess[] = [];
+
+  constructor() {
+    this.maxParallel = os.cpus().length;
+  }
+
+  setMaxParallel(n) {
+    this.maxParallel = n;
+  }
+
+  async getOrCreate(isTypescriptRunMode): Promise<IMochaProcess> {
+    const extension = isTypescriptRunMode ? 'ts' : 'js';
+    const runnerPath = pathResolve(__dirname, `../subprocess/runner.${extension}`);
+
+    const lastUnusedProcess = this.unusedProcesses.pop();
+    if (lastUnusedProcess) {
+      return lastUnusedProcess;
+    }
+
+    if (this.processes.length >= this.maxParallel) {
+      const process: IMochaProcess = await new Promise<IMochaProcess>((resolve) => {
+        this.waitingList.push((proc: IMochaProcess) => {
+          resolve(proc);
+        });
+      });
+      return process;
+    }
+    return this.create(runnerPath, {
+      // otherwise `--inspect-brk` and other params will be passed to subprocess
+      execArgv: process.execArgv.filter(removeDebugArgs),
+      stdio: ['ipc'],
+    });
+  }
+
+  create(runnerPath, opt) {
+    const process = fork(runnerPath, [], opt) as IMochaProcess;
+
+    this.processes.push(process);
+
+    process.destroy = () => {
```

Comment:
I don't like the idea of changing native (node.js) objects: it may work now but it can break afterwards, the object can become frozen etc. A similar yet not exactly the same thing is the "Can you add methods to Array.prototype" discussion: the idea is pretty close. I left a comment somewhere closer to the top of this PR if you would like to consider using your own abstraction (say, a simple class with a "send" and "destroy" methods which will pass these commands to a forked process which will be a private property). I guess it's worth it to think about this thing.